refactored and improved minimal perfect hash code (regen/mph.pl)

demerphq opened the pull request:
Some time back I wrote regen/mph.pl for @khwilliamson to use in parsing our unicode properties. I had completely forgotten this code and rediscovered it when debugging https://github.com/Perl/perl5/issues/19603 (I wrote it at a time when I was under great stress after my mom passed away).

I noticed some odd things when it ran and decided to clean it up, fix the issues I saw and generally reacquaint myself with the code. While digging into it I realized it was missing certain optimizations and ran very very slowly. The version of regen/mk_invlists.pl in blead right now takes about 11.5 minutes to run on my laptop, about 9.5 of that coming from regen/mph.pl

I fixed the major of the cause of the slowness by removing "use bignum" which had been added to make it work on 32 bit perls, and then added some optimizations and improvements. The new version of regen/mk_invlists runs in about 3.5 minutes, with about 1 minute due to mph.pl and at the same time it produces a significantly (~20%) smaller blob for the keys.

Some context: mph.pl constructs a two level minimal perfect hash, and it constructs a minimal covering string of split keys. The two level minimal perfect hash algorithm involves building a special type of hash tale with exactly the number of buckets as there are keys.  The minimal covering set of split keys is a way of compressing the keys such that they are represented as a prefix and a suffix. For our use case (And many use cases) it turns out that by splitting the keys we can save a lot of space by reusing the prefixes and suffixes or parts of them, for multiple keys. 

This PR includes several improvements to both algorithms. 

The improvements to the minimal perfect hash construction means it should execute in nearly O(N) time, and the improvements to the minimal covering string algorithm mean it produces a smaller buffer to represent the keys. 

The algorithm for the minimal covering string algorithm "_squeeze" is by a former colleague of mine Ilya Sashcheka who I have unfortunately lost contact with and no longer have an email for, but I have added him to the AUTHORS file none the less. He gave me permission to use this code back when we worked together but it just never made it into core until now. 

The combination of the two algorithms means we can represent the entire data set in about 65% of the size of the raw keys themselves. This is substantially smaller than the size of code we would produce if we used Devel::Tokenizer:C which would be much larger than the 104237 bytes of raw keys we need to represent.

The net result of this patch is that it takes about 3.5 minutes to run regen/mk_invlists.pl and it produces a key buffer that is 8635 verus the current 10617 bytes. IOW, it saves about 7 minutes and 2k of ram.  As a bonus this version of the code is easier to work on and maintain as I have converted it to use OO for the public API, and has way more comments about how it works.

Included in this PR is a fix for t/op/magic.t that i noticed while working on the PR.

demerphq committed:
regen/mph.pl - document build_split_words()

Explain the behavior and inputs and outputs of build_split_words()

demerphq committed:
regen/mph.pl - refactor common sort routine out

This sort expression gets repeated a lot and it is quite long, replace
with a sub.

demerphq committed:
regen/mph.pl - remove unused arguments from build_split_words()

These were vestigal from a previous implementation, no point in
leaving them in, and they impact debugging.

As part of this it make sense to rename %res to $res in
build_scalar_words() as it makes it easier to make $old_res= $res
when we retry.

demerphq committed:
regen/mph.pl - dont use $key when dealing with $part of a key

In build_split_words() $key is used for the main hash, use
$part instead when we are dealing with part of a key (even
if the $part might equal the $key).

demerphq committed:
regen/mph.pl - dont null terminate in preprocess stage

Removing the null termination allows us to save about 100 bytes
from the final result.

demerphq committed:
regen/mph.pl - track added bytes while we process/check the blob

It is possible that the $blob grows while we validate it, if so
show how much it grew by and produce some diagnostics about it.
We also track how many passes we have done. With this commit it
is only used to make the new diagnostics a bit cleaner, but we will
use it in more diagnostics later.

demerphq committed:
regen/mph.pl - do a bit less work in build_split_words()

We do not need to check the split point of 0 or of the length of the
string as we do not want empty prefixes (the 0 case) and we have already
checked if the entire string is in the blob already in a earlier check
(the length case). This also initializes the $best_prefix to be the full
$key, and the $best_suffix to be the empty string just in case we cannot
find any split point which would result in the variables being
initialized. This prevents unitialized warnings when we track these
variables in the %appended hash.

demerphq committed:
regen/mph.pl - fix assignment to be consistent with other assignments.

This file uses left hugging assignment operators, this was an exception.

demerphq committed:
regen/mph.pl - better diags in build_split_words()

This also shows how effective the split process compression has been
by computing what percentage the final blob is compared to the worst
case of naively concatenating all the keys together.

demerphq committed:
regen/mph.pl - rename var $b2 to $new_blob in build_split_words()

$new_blob is more descriptive and less confusing.

demerphq committed:
regen/mph.pl - calculate $length_all_keys in make_mph_from_hash()

Doing it in build_perfect_hash() just results in duplicated logic
between build_split_words() and build_perfect_hash()

demerphq committed:
regen/mph.pl - handle split data in make_mph_from_hash()

Doing it in build_perfect_hash() does not make sense, we might want to
use that function to build a hash that does not need to be split.

demerphq committed:
regen/mk_invlists.pl - move require to top of file

mk_invlists.pl does a lot and takes a while before it gets to the part
where it requires regen/mph.pl, which means that if there are issues in
it they arent discovered until a fair amount of time elapses, which is
frustrating when debugging. Moving the require to the top means the
script dies early and can be fixed.

Includes a regen of uni_keywords.h and friends as this changes a regen
script which causes regen.t to fail if its output is not up to date.

demerphq committed:
regen/mk_invlists.pl - move token_name() sub closer to where it is used

sub token_name() was injected into the middle of totally unrelated logic
that does not use it. token_name() is a wrapper around sanitize_name()
so move it next to that sub.

Also includes the output from running regen/mk_invlists.pl to keep
porting/regen.t happy.

demerphq committed:
regen/mk_invlists.pl - add a way to dump the keywords hash for review

This adds a way to tell mk_invlists.pl to dump the keywords hash so it
can be reviewed, or used for testing or whatnot. A user can define the
env var DUMP_KEYWORDS_FILE to be a file name which will be used to save
the keywords hash to. If the env var is not set the file won't get
written to disk.

Includes regenerated output from running regen/mk_invlists.pl to keep
porting/regen.t happy.

demerphq committed:
regen/mph.pl - remove unnecessary use of bignum

We can detect we are on a 32 bit perl and then do the right thing. This
*massively* speeds up the process. Hashing character by character with
bignum enabled is painfully slow.

demerphq committed:
regen/mph.pl - replace hard coded literal constants

This replaces the various hard coded constants with either named
constants, or vars. It also renames FNV_CONST to FNV32_PRIME in the
generated code as that is its proper name.

demerphq committed:
regen/mph.pl - specify which fnv we are using

This documents and specifies which FNV we are using. Since we are using
fnv1a_32() it also replaces the use of $MASK and replaces it with
U32_MAX. fnv1a_32() would never use anything other U32_MAX, even if
$MASK changed, as obviously it is a 32 bit hash function.

demerphq committed:
regen/mph.pl - remove $max_h check in build_perfect_hash()

Not sure why I put this in to the original code, it is a check that
would be relevant to a PRNG but is not relevant to this use case. So
this patch removes it.

demerphq committed:
regen/mph.pl - perltidy file for style consistency

and document the perltidy options used so any future maintainers
can follow the style of the file more easily.

demerphq committed:
regen/mph.pl - remove unused var

demerphq committed:
regen/mph.pl - enable fatal warnings

If this script warns then there is something very wrong and it should be
fixed, so just die immediately. Especially as if it is broken it might
just *spew* warnings, which is annoying.

demerphq committed:
regen/mph.pl - convert loop to use block form and add comment

Using the comma operator to separate statments is fine on a one liner,
but so much in a script that is part of perl regen processes. IMO.

demerphq committed:
regen/mph.pl - add sanity check for idx and value parameters

If either are missing then something has gone badly wrong and we should
stop processing immediately.

demerphq committed:
regen/mph.pl - bucket info storage: remove 'hash', move 'value' logic

The 'hash' key is totally unused and unneeded so drop it entirely.

The 'value' key can be stored into the bucket info data elsewhere,
strictly speaking it is not needed for the minimal perfect hash
computations, so move it out so that logic can be changed and
simplified in a future patch.

demerphq committed:
regen/mph.pl - move bucket info construction log into a sub

In a follow up patch this logic will get called from more than one place
so move it to a sub so it can be used easily.

demerphq committed:
regen/mph.pl - change $hash argument to $source_hash

The term 'hash' is overused in this code, rename the $hash argument
to build_perfect_hash() to $source_hash to disambiguate.

demerphq committed:
regen/mph.pl - eliminate the need to use goto

The goto is confusing, and has the potential to introduce its own bugs
if future changes are not careful, so get rid of it completely and
break build_perfect_hash() into two subs.

demerphq committed:
regen/mph.pl - use more efficient logic in build_perfect_hash()

This patch greatly reduces the amount of work we have to do to find a
bucket for keys whose first level hash does not produce any collisions.

We have two cases, those items whose $hash2 is higher than $MAX_SEED2
and those items whose $hash2 is smaller than or equal to $MAX_SEED2.

For the former case we use a similar but streamlined process as we do
for keys whose first level hash produces collisions. For the latter case
we can trivially map the items into any bucket we choose.

demerphq committed:
regen/mph.pl - Clean up diagnostics logic, allow DEBUG from env.

Be silent unless requested to. If DEBUG>1 produce lots of output,
if DEBUG==1 produce some basic information about what is going on.

demerphq committed:
op/magic.t - make $SIG{ALRM} local test deal with inherited IGNORE'd signal

Signal ignores can be set up by a parent process and the child process
will inherit them. So in some situations when we test $SIG{ALRM} might
be "IGNORE" and not undef. So instead of just expecting undef, check
what it was before the localization, and then check it is still that
value afterwards.

I found this while running make test via rebase --exec, something like
this reduction (from Matthew Horsfall):

    $ git rebase --exec='perl -le"print \$SIG{ALRM}"' HEAD~1
    IGNORE

A similar case for a different signal would be this example (from Leon
Timmermans):

    $ nohup perl -E 'say $SIG{HUP}' 2>/dev/null | cat
    IGNORE

demerphq committed:
regen/mph.pl - fixup die that is issued if we can't solve this hash

The old code was bit messed up, but I didn't notice because it doesn't
happen with our current data. But in theory it could. It is possible
that fnv1a has a multicollision vulnerability as it is not a secure
hash, so changing the seed wouldn't help. For now we can assume it does
not.

demerphq committed:
regen/mph.pl - change fnv1a_32() to _fnv1a_32() as it is not public

This function is not part of the "public" API for this script/package
and might be removed in the future if we needed to, so mark it as
private with a leading underscore.

demerphq committed:
regen/mph.pl - fixup wording in comment to be grammatical

The original wording of this paragraph was a bit clumsy and not
grammatically correct. This fixes it.

demerphq committed:
regen/mph.pl - move split key logic out of make_mph_from_hash

The logic of calling build_split_words() twice, once in "normal" mode
and once in "preproces" mode does not belong in regen/mph.pl. So this
patch renames build_split_words() to _build_split_words() and create
a new sub called build_split_words() that implements the "call it
twice" logic.

For consistency the arguments are rearranged as well so the $preprocess
argument is last, as build_split_words() now does not have a $preproces
argument, as only _build_split_words() needs it.

demerphq committed:
regen/mph.pl - rename $smart_blob to $blob across file

$smart_blob is unhelpfully specific, rename to more generic $blob

demerphq committed:
regen/mph.pl - whitespace fixups

This removes trailing whitespace only. No logic changes here at all.

demerphq committed:
regen/mph.pl - move file open logic to a sub

this makes all the subs affected support a filehandle or filename

demerphq committed:
regen/mph.pl - remove bogus defaulting for undef vars

This was some kind of thinko and never made sense, prefix
and suffix should never be undefined.

demerphq committed:
regen/mph.pl - use $n instead of @$second_level

Same thing but $n is initialized one line above, so use it.

demerphq committed:
regen/mph.pl - tweaks to generated code, put type on its own line

All the C code we have puts the type on its own line separate from
the function parameter declaration, so follow that style in our
generated file too.

Also show the generator script in the comment that contains metadata
about the file.

demerphq committed:
regen/mph.pl & mk_invlists.pl - convert from sub interfaces to OO interfaces

The old sub based API was passing around an awkward number of arguments
and it was becoming difficult to enhance in certain ways. This patch
changes all the "user servicable" functions into methods, and moves the
configuration defaults into the constructor.

Note, not all the functions have been converted, the core routines with
simple interfaces have not been changed. This is OO for the purpose of
encapsulation not inheritance or overloading.

demerphq committed:
regen/mph.pl & mk_invlists.pl - add the "_squeeze" algorithm to produce smaller blobs

The squeeze algorithm produces smaller blobs, 10-20% depending on how it
is used. With the "randomize_squeeze" option enabled it is slower but
produces 20% smaller blobs than the "_simple" strategy we used to use.
With the "randomize_squeeze" option disabled it is about as fast as
"_simple" but produces about 10% smaller blobs. Regardless "_squeeze"
uses more memory than _simple; quite a bit more currently, although that
is unforced and could be changed if required.

    -blob length: 10548
    +blob length: 8635
    ...
    -data size: 69908 (%67.07)
    +data size: 67995 (%65.23)

So it saves 1913 bytes running with this seed. I happened to get lucky
with the seed, depending on the seed used the blob ended up about 8650
bytes.

This algorithm is originally by Ilya Sashcheka, so I have added him to
the AUTHORS file, but unfortunately I no longer have his email address
as we lost touch. It contains many modifications by me.

haarg commented on regen/mph.pl:
```diff
@@ -32,160 +112,382 @@ package MinimalPerfectHash;
+    #
+    # Each type of bucket uses a different algorithm to solve. Note that
+    # a "classical" two level hash would only have "hard" and "trivial"
+    # buckets, but since we supporint having a larger hash value than we
```
typo

demerphq replied:
Fixed.

khwilliamson reviewed (changes requested):
LGTM, but it needs to regen mk_invlist to not fail porting tests.  The CI won' run without that fix.  I did it on my box locally, and all tests passed.

demerphq committed:
regen/mph.pl - add a validation step to build_split_words()

Exercise an abundance of caution and validate that the buffer and split
point data returned is fit for pupose.

Includes the output of running regen/mk_invlist.pl.

demerphq commented:
I modified the old last commit to include the mk_invlist.pl changes, and I added an additional sanity check patch that validates the split keys buffer and regenerated again.
